modules/core/include/opencv2/core/mat.hpp

```diff
@@ -2451,11 +2451,11 @@ class CV_EXPORTS UMat
     //!  <0 - a diagonal from the lower half)
     UMat diag(int d=0) const;
     //! constructs a square diagonal matrix which main diagonal is vector "d"
-    static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-    static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
+    CV_NODISCARD_STD static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
+    CV_NODISCARD_STD static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
```

Comment:
Resolved `CV_NODISCARD_STD` conflicts from patch #20622

/cc @diablodale 

BTW, why nodiscard flag is missing for methods above, like

`UMat diag(int d=0) const;`

Reply:
Not having CV_NODISCARD_STD on `UMat diag(int d=0)` was a style choice. `diag` is a member function, not a static factory. Static factories have the "feature" that they can be called using `class::factory` or `var.factory` which contributes to discarding errors.

There are hundreds/thousands of APIs in opencv that create an object and return it. Technically... nodiscard could be put on all of them. But that is overkill and not really the intention of [[nodiscard]]. For this first PR effort, I focused on the static factories that create new independent objects. There may be future opportunities on some constructors (like those that do locking), but that wasn't the focus on this first effort. For example, I could put nodicard on the UMat constructors, but it is harder to make a discarding error with UMat constructor...compared to a discarding error with UMat::zeros

The one exception was member functions `clone()`. Those were already marked and therefore I want to maintain the same functionality on them.

Reply:
Also... I don't visually 👀 see any differences between my original nodiscard commit and this PR's changes.
You wrote you resolved conflicts. Is there a way in github to see the specific changes you applied to resolve conflicts?

Reply:
this block merge changes looks ok also

Reply:
`git show <merge commit sha>` can show conflicting changes:

<details>

```.diff
commit 5aa7435d2537e677611390b530e31b3b204773a3 (HEAD -> merge-3.4, github/test_contrib, github/merge-3.4)
Merge: c11195d5e3 7c23ec90a9
Author: Alexander Alekhin <alexander.a.alekhin@gmail.com>
Date:   Thu Sep 2 15:24:04 2021 +0000

    Merge remote-tracking branch 'upstream/3.4' into merge-3.4

diff --cc modules/core/include/opencv2/core/mat.hpp
index 6768be7683,fdcb8fc817..1ef0eb5a02
--- a/modules/core/include/opencv2/core/mat.hpp
+++ b/modules/core/include/opencv2/core/mat.hpp
@@@ -2451,11 -2469,10 +2451,11 @@@ public
      //!  <0 - a diagonal from the lower half)
      UMat diag(int d=0) const;
      //! constructs a square diagonal matrix which main diagonal is vector "d"
-     static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
 -    CV_NODISCARD_STD static UMat diag(const UMat& d);
++    CV_NODISCARD_STD static UMat diag(const UMat& d, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat diag(const UMat& d) { return diag(d, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
  
      //! returns deep copy of the matrix, i.e. the data is copied
-     UMat clone() const CV_NODISCARD;
+     CV_NODISCARD_STD UMat clone() const;
      //! copies the matrix content to "m".
      // It calls m.create(this->size(), this->type()).
      void copyTo( OutputArray m ) const;
@@@ -2486,22 -2503,14 +2486,22 @@@
      double dot(InputArray m) const;
  
      //! Matlab-style matrix initialization
-     static UMat zeros(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat zeros(int rows, int cols, int type) { return zeros(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat zeros(Size size, int type) { return zeros(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat zeros(int ndims, const int* sz, int type) { return zeros(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat ones(int rows, int cols, int type) { return ones(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(Size size, int type) { return ones(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat ones(int ndims, const int* sz, int type) { return ones(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat eye(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat eye(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
-     static UMat eye(int rows, int cols, int type) { return eye(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
-     static UMat eye(Size size, int type) { return eye(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
 -    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat zeros(Size size, int type);
 -    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type);
 -    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat ones(Size size, int type);
 -    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type);
 -    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type);
 -    CV_NODISCARD_STD static UMat eye(Size size, int type);
++    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat zeros(int rows, int cols, int type) { return zeros(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat zeros(Size size, int type) { return zeros(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat zeros(int ndims, const int* sz, int type) { return zeros(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat ones(int rows, int cols, int type) { return ones(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(Size size, int type) { return ones(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat ones(int ndims, const int* sz, int type) { return ones(ndims, sz, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat eye(Size size, int type, UMatUsageFlags usageFlags /*= USAGE_DEFAULT*/);
++    CV_NODISCARD_STD static UMat eye(int rows, int cols, int type) { return eye(rows, cols, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
++    CV_NODISCARD_STD static UMat eye(Size size, int type) { return eye(size, type, USAGE_DEFAULT); }  // OpenCV 5.0: remove abi compatibility overload
  
      //! allocates new matrix data unless the matrix already has specified size and type.
      // previous data is unreferenced if needed.
diff --cc modules/core/include/opencv2/core/matx.hpp
index 45d8a97f34,f25c8bce57..3c92e3a21d
--- a/modules/core/include/opencv2/core/matx.hpp
+++ b/modules/core/include/opencv2/core/matx.hpp
@@@ -140,13 -142,15 +140,13 @@@ public
           _Tp v12, _Tp v13, _Tp v14, _Tp v15); //!< 1x16, 4x4 or 16x1 matrix
      explicit Matx(const _Tp* vals); //!< initialize from a plain array
  
 -#ifdef CV_CXX11
      Matx(std::initializer_list<_Tp>); //!< initialize from an initializer list
 -#endif
  
-     static Matx all(_Tp alpha);
-     static Matx zeros();
-     static Matx ones();
-     static Matx eye();
-     static Matx diag(const diag_type& d);
+     CV_NODISCARD_STD static Matx all(_Tp alpha);
+     CV_NODISCARD_STD static Matx zeros();
+     CV_NODISCARD_STD static Matx ones();
+     CV_NODISCARD_STD static Matx eye();
+     CV_NODISCARD_STD static Matx diag(const diag_type& d);
      /** @brief Generates uniformly distributed random numbers
      @param a Range boundary.
      @param b The other range boundary (boundaries don't have to be ordered, the lower boundary is inclusive,
diff --cc modules/core/src/bindings_utils.cpp
...
```

</details>